addon/components/collection-scroll-view/index.js

```diff
@@ -47,96 +45,223 @@ export default class CollectionScrollView extends Component {
     if (
       source &&
-      (isChildComponent(this, source) || isChildElement(this.element, source))
+      (isChildComponent(this, source) ||
+        isChildElement(this.scrollElement, source))
     ) {
       return;
     }
-    let { items } = this.args;
-    let itemIndex = items.indexOf(items.find((i) => i.id === id));
-    if (itemIndex >= 0) {
-      let { y } = this.cellLayout.positionAt(itemIndex);
-      scrollViewApi.scrollTo(y + this.headerHeight, true);
+    if (!this.verticalCollectionApi) {
+      return;
+    }
+    await this.verticalCollectionApi.scrollToItem(id);
+  };
+
+  handleResize = (entry) => {
+    let rect = entry?.contentRect;
+    let width;
+    let height;
+
+    if (rect) {
+      ({ width, height } = rect);
+    } else if (this.scrollElement) {
+      width = this.scrollElement.clientWidth;
+      height = this.scrollElement.clientHeight;
+    }
+
+    if (width !== undefined) {
+      this.clientWidth = width;
+    }
+
+    let previousHeight = this.clientHeight;
+    if (height !== undefined) {
+      this.clientHeight = height;
+    }
+
+    if (height !== undefined && height !== previousHeight) {
+      this.scheduleVerticalCollectionRefresh();
+    }
+
+    this.schedulePendingRestore();
+  };
+
+  scheduleVerticalCollectionRefresh() {
+    if (this._verticalCollectionRefreshScheduled) {
+      return;
+    }
+    this._verticalCollectionRefreshScheduled = true;
+    scheduleOnce('afterRender', this, this.dispatchVerticalCollectionRefresh);
+  }
+
+  dispatchVerticalCollectionRefresh() {
+    this._verticalCollectionRefreshScheduled = false;
+    if (typeof window === 'undefined') {
+      return;
+    }
+    window.dispatchEvent(new Event('resize'));
+  }
+
+  storeScrollPosition() {
+    if (!this.args.memoryScrollKey || !this.scrollElement) {
+      return;
+    }
+    let scrollTop = this.scrollElement.scrollTop ?? 0;
+    this.memory[this.args.memoryScrollKey] = scrollTop;
+  }
+
+  restoreScrollPosition() {
+    const { initialScrollTop, memoryScrollKey } = this.args;
+    if (!memoryScrollKey && !initialScrollTop) {
+      return;
+    }
+    let storedPosition = this.memory[memoryScrollKey];
+    if (typeof storedPosition !== 'number') {
+      storedPosition = initialScrollTop;
+    }
+    if (typeof storedPosition !== 'number') {
+      return;
+    }
+    this._pendingRestorePosition = storedPosition;
+    this.schedulePendingRestore();
+  }
+
+  schedulePendingRestore() {
+    if (this._pendingRestorePosition === undefined || this._restoreScheduled) {
+      return;
+    }
+    this._restoreScheduled = true;
+    scheduleOnce('afterRender', this, this.applyPendingRestore);
+  }
+
+  applyPendingRestore() {
+    this._restoreScheduled = false;
+    let position = this._pendingRestorePosition;
+    if (position === undefined) {
+      return;
     }
+    if (!this.canRestoreScrollPosition()) {
+      this.schedulePendingRestore();
+      return;
+    }
```

Comment:
Consider a max number of attempts for this to avoid an endless or very long loop./